lib/elixir/lib/kernel.ex

```diff
@@ -2198,6 +2198,54 @@ defmodule Kernel do
     :erlang.error(ArgumentError.exception(error_message))
   end
 
+  @doc """
+  Returns true if `term` is a structure; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term) do
+    case __CALLER__.context do
+      nil ->
```

Comment:
This clause can be the same as the `:guard` one. We just need to check we are not inside a match.

Reply:
Actually, ignore this comment for now.

Reply:
It would need to be `is_map(term) and :erlang.map_get(key, term)` to prevent errors on non-map `term`

Reply:
> This clause can be the same as the `:guard` one. We just need to check we are not inside a match.

if that's the case, wouldn't it be better off to inline this function?

Reply:
> It would need to be is_map(term) and :erlang.map_get(key, term) to prevent errors on non-map term

Yeah, we also need this in the other one. But this code is simpler to emit and verify for the non-guard case, so let's please keep this one as is.

Reply:
> @josevalim if that's the case, wouldn't it be better off to inline this function?

We only inline when there is a direct equivalent in Erlang. There isn't one here.

Reply:
> > @josevalim if that's the case, wouldn't it be better off to inline this function?
> 
> We only inline when there is a direct equivalent in Erlang. There isn't one here.

Oh! good to know